File: apps/web/src/api/wiki.ts
import apiClient from "./client";

export interface WikiQualityScore {
  source_coverage: number;
  freshness: number;
  consistency: number;
  connectivity: number;
  overall: number;
  supported_claims: number;
  unsupported_claims: number;
  open_conflicts: number;
  connections_count: number;
  issues: string[];
}

export interface WikiPageItem {
  id: string;
  document_id: string | null;
  slug: string;
  title: string;
  summary: string;
  content_markdown: string;
  created_at: string;
  updated_at: string;
  quality: WikiQualityScore | null;
}

export interface WikiPageSourceItem {
  chunk_id: string;
  document_id: string;
  document_filename: string;
  chunk_index: number;
  page_number: number | null;
}

export interface WikiClaimCitationItem {
  claim_key: string;
  claim_text: string;
  position: number;
  sources: WikiPageSourceItem[];
}

export interface WikiPageReferenceItem {
  page_id: string;
  slug: string;
  title: string;
  label: string;
}

export interface WikiPageConflictItem {
  id: string;
  source_document_id: string | null;
  existing_statement: string;
  incoming_statement: string;
  explanation: string;
  status: "OPEN" | "RESOLVED" | "DISMISSED";
  resolution_note: string;
  created_at: string;
  resolved_at: string | null;
}

export interface WikiPageDetails extends WikiPageItem {
  sources: WikiPageSourceItem[];
  related_pages: WikiPageReferenceItem[];
  backlinks: WikiPageReferenceItem[];
  conflicts: WikiPageConflictItem[];
  claim_citations: WikiClaimCitationItem[];
}

export interface WikiPageRevisionItem {
  id: string;
  wiki_page_id: string | null;
  page_slug: string;
  revision_number: number;
  title: string;
  summary: string;
  content_markdown: string;
  operation: "CREATE" | "UPDATE" | "MERGE" | "RESTORE";
  triggering_document_id: string | null;
  created_at: string;
}

export interface WikiRevisionDiffLine {
  kind: "added" | "removed" | "context";
  text: string;
}

export interface WikiRevisionDiff {
  from_revision_number: number | null;
  to_revision_number: number;
  lines: WikiRevisionDiffLine[];
}

export type WikiMaintenanceStatus =
  | "PENDING"
  | "APPROVED"
  | "REJECTED";

export type WikiMaintenanceIssueType =
  | "DUPLICATE"
  | "ORPHAN"
  | "BROKEN_LINK"
  | "OVERSIZED"
  | "UNDERSIZED"
  | "UNSUPPORTED_CLAIMS";

export interface WikiMaintenanceSuggestion {
  id: string;
  issue_type: WikiMaintenanceIssueType;
  status: WikiMaintenanceStatus;
  title: string;
  description: string;
  page_ids: string[];
  metadata: Record<string, unknown>;
  confidence: number;
  created_at: string;
  updated_at: string;
}

export interface CompileWikiResponse {
  document_id: string;
  pages_count: number;
  sources_count: number;
  links_count: number;
  conflicts_count: number;
  pages: WikiPageItem[];
}

export async function compileDocumentWiki(
  documentId: string,
): Promise<CompileWikiResponse> {
  const response = await apiClient.post<CompileWikiResponse>(
    `/wiki/documents/${documentId}/compile`,
  );

  return response.data;
}

export async function listWikiPages(): Promise<WikiPageItem[]> {
  const response = await apiClient.get<WikiPageItem[]>("/wiki/pages");
  return response.data;
}

export async function getWikiPage(
  slug: string,
): Promise<WikiPageDetails> {
  const response = await apiClient.get<WikiPageDetails>(
    `/wiki/pages/${encodeURIComponent(slug)}`,
  );

  return response.data;
}

export async function listWikiPageRevisions(
  slug: string,
): Promise<WikiPageRevisionItem[]> {
  const response = await apiClient.get<WikiPageRevisionItem[]>(
    `/wiki/pages/${encodeURIComponent(slug)}/revisions`,
  );

  return response.data;
}

export async function getWikiRevisionDiff(
  slug: string,
  revisionNumber: number,
): Promise<WikiRevisionDiff> {
  const response = await apiClient.get<WikiRevisionDiff>(
    `/wiki/pages/${encodeURIComponent(
      slug,
    )}/revisions/${revisionNumber}/diff`,
  );

  return response.data;
}

export async function restoreWikiRevision(
  slug: string,
  revisionNumber: number,
): Promise<WikiPageItem> {
  const response = await apiClient.post<WikiPageItem>(
    `/wiki/pages/${encodeURIComponent(
      slug,
    )}/revisions/${revisionNumber}/restore`,
  );

  return response.data;
}

export async function updateWikiConflict(
  conflictId: string,
  status: "OPEN" | "RESOLVED" | "DISMISSED",
  resolutionNote = "",
): Promise<WikiPageConflictItem> {
  const response = await apiClient.patch<WikiPageConflictItem>(
    `/wiki/conflicts/${conflictId}`,
    {
      status,
      resolution_note: resolutionNote,
    },
  );

  return response.data;
}

export async function scanWikiMaintenance(): Promise<
  WikiMaintenanceSuggestion[]
> {
  const response = await apiClient.post<WikiMaintenanceSuggestion[]>(
    "/wiki/maintenance/scan",
  );
  return response.data;
}

export async function listWikiMaintenanceSuggestions(): Promise<
  WikiMaintenanceSuggestion[]
> {
  const response = await apiClient.get<WikiMaintenanceSuggestion[]>(
    "/wiki/maintenance/suggestions",
  );
  return response.data;
}

export async function updateWikiMaintenanceSuggestion(
  suggestionId: string,
  status: "APPROVED" | "REJECTED",
): Promise<WikiMaintenanceSuggestion> {
  const response = await apiClient.patch<WikiMaintenanceSuggestion>(
    `/wiki/maintenance/suggestions/${suggestionId}`,
    { status },
  );
  return response.data;
}

File: apps/api/src/knowledge_assistant/presentation/api/v1/schemas/wiki.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class WikiQualityScoreResponse(BaseModel):
    source_coverage: int
    freshness: int
    consistency: int
    connectivity: int
    overall: int
    supported_claims: int
    unsupported_claims: int
    open_conflicts: int
    connections_count: int
    issues: list[str]


class WikiPageResponse(BaseModel):
    id: UUID
    document_id: UUID | None
    slug: str
    title: str
    summary: str
    content_markdown: str
    created_at: datetime
    updated_at: datetime
    quality: WikiQualityScoreResponse | None = None


class WikiPageSourceResponse(BaseModel):
    chunk_id: UUID
    document_id: UUID
    document_filename: str
    chunk_index: int
    page_number: int | None


class WikiClaimCitationResponse(BaseModel):
    claim_key: str
    claim_text: str
    position: int
    sources: list[WikiPageSourceResponse]


class WikiPageReferenceResponse(BaseModel):
    page_id: UUID
    slug: str
    title: str
    label: str


class WikiPageConflictResponse(BaseModel):
    id: UUID
    source_document_id: UUID | None
    existing_statement: str
    incoming_statement: str
    explanation: str
    status: str
    resolution_note: str
    created_at: datetime
    resolved_at: datetime | None


class WikiPageDetailsResponse(WikiPageResponse):
    sources: list[WikiPageSourceResponse]
    related_pages: list[WikiPageReferenceResponse]
    backlinks: list[WikiPageReferenceResponse]
    conflicts: list[WikiPageConflictResponse]
    claim_citations: list[WikiClaimCitationResponse]


class WikiPageRevisionResponse(BaseModel):
    id: UUID
    wiki_page_id: UUID | None
    page_slug: str
    revision_number: int
    title: str
    summary: str
    content_markdown: str
    operation: str
    triggering_document_id: UUID | None
    created_at: datetime


class WikiRevisionDiffLineResponse(BaseModel):
    kind: Literal["added", "removed", "context"]
    text: str


class WikiRevisionDiffResponse(BaseModel):
    from_revision_number: int | None
    to_revision_number: int
    lines: list[WikiRevisionDiffLineResponse]


class UpdateWikiConflictRequest(BaseModel):
    status: Literal["OPEN", "RESOLVED", "DISMISSED"]
    resolution_note: str = Field(default="", max_length=2000)


class CompileWikiResponse(BaseModel):
    document_id: UUID
    pages_count: int
    sources_count: int
    links_count: int
    conflicts_count: int
    pages: list[WikiPageResponse]


class WikiMaintenanceSuggestionResponse(BaseModel):
    id: UUID
    issue_type: str
    status: str
    title: str
    description: str
    page_ids: list[UUID]
    metadata: dict[str, Any]
    confidence: float
    created_at: datetime
    updated_at: datetime


class UpdateWikiMaintenanceSuggestionRequest(BaseModel):
    status: Literal["APPROVED", "REJECTED"]
